### PR TITLE
Implement single-file "find references" in LSP server

### PR DESCRIPTION
**Pull request by Garmelon**
This implementation will only find occurrences in the current file (even if the declaration is in another file). Works for global definitions and local variables.

The relevant issue #523 should probably not yet be closed because of the single-file limitation.

I still need to add a few tests.

### CODE REVIEW

**Comment by Garmelon** [comment]
At the moment, this PR is waiting for a solution to #840

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 1]
```diff
@@ -134,6 +134,102 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
             return ← ci.runMetaM i.lctx <| locationLinksFromDecl i ei.elaborator
       return #[]
 
+inductive RefIdent where
+  | name : Name → RefIdent
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:138]: Let's call this one `const` for consistency

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 2]
```diff
@@ -134,6 +134,102 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
...
+        | Expr.const n .. => some (RefIdent.name n, ti.isBinder)
+        | Expr.fvar id .. => some (RefIdent.fvar id, ti.isBinder)
+        | _ => none
+      | Info.ofFieldInfo fi => some (RefIdent.name fi.projName, false)
+      | _ => none
+
+    addReference (doc : EditableDocument) (info : Info)
+        (ident : RefIdent) (isDeclaration : Bool)
+        (refs : Array Reference) : Array Reference :=
+      if let some range := info.stx.getRange? (originalOnly := true) then
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:174]:
```suggestion
      if let some range := info.range? then
```
same meaning

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 3]
```diff
@@ -134,6 +134,102 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
...
+        let text := doc.meta.text
+        let range := {
+          start := text.utf8PosToLspPos range.start
+          «end» := text.utf8PosToLspPos range.stop
+        }
+        refs.push { ident, range, isDeclaration }
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:180]:
```suggestion
        refs.push { ident, range := range.toLspRange text, isDeclaration }
```

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 4]
```diff
@@ -134,6 +134,102 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
...
+      else
+        refs
+
+    addReferences (doc : EditableDocument) (snap : Snapshot)
+        (refs : Array Reference) : Array Reference :=
+      let infos := snap.infoTree.findAll fun
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:186]: Can we use `deepestNodes` here? E.g. `snap.infoTree.deepestNodes (fun _ i _ => identOf i)`.

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 5]
```diff
@@ -134,6 +134,102 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
...
+        | Info.ofTermInfo _ => true
+        | Info.ofFieldInfo _ => true
+        | _ => false
+      infos.foldl (init := refs) fun refs info => do
+        if let some (ident, isDecl) := identOf info then
+          return addReference doc info ident isDecl refs
+        return refs
+
+    findReferences (doc : EditableDocument) (snaps : List Snapshot) : Array Reference :=
+      snaps.foldl (init := #[]) fun refs snap => addReferences doc snap refs
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:196]: You might consider inlining `addReferences` and `addReference` into a big loop(s) with a mutable variable to get rid of the many parameters, but this way is also fine

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 6]
```diff
@@ -134,6 +134,102 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
...
+
+    applyIdMap : RefIdent → HashMap FVarId FVarId → RefIdent
+      | RefIdent.fvar id, m => RefIdent.fvar <| m.findD id id
+      | ident, _ => ident
+
+    -- The `FVarId`s of a function parameter in the function's definition and
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:202]:
```suggestion
    -- The `FVarId`s of a function parameter in the function's signature and
```

File: src/Lean/Server/FileWorker/RequestHandling.lean [hunk 7]
```diff
@@ -134,6 +134,102 @@ partial def handleDefinition (kind : GoToKind) (p : TextDocumentPositionParams)
...
+          refs.push { ident := applyIdMap ident idMap, range, isDeclaration := false }
+        | _ => refs.push ref
+      (refs, idMap)
+
+    referencesTo (doc : EditableDocument) (ident : RefIdent) (includeDecl : Bool) (snaps : List Snapshot)
+        : Array Location :=
+      let (refs, idMap) := combineFvars $ findReferences doc snaps
+      let ident := applyIdMap ident idMap
+      let relevant := refs.filter fun ref => ref.ident == ident && (includeDecl || !ref.isDeclaration)
+      let locations := relevant.map fun ref => { uri := doc.meta.uri, range := ref.range : Location}
```
Comment (Kha) [inline src/Lean/Server/FileWorker/RequestHandling.lean:229]:
```suggestion
      let locations := relevant.map fun ref => { uri := doc.meta.uri, range := ref.range : Location }
```

**Review by Kha — APPROVED** [review]

**Pull request merged**